### PR TITLE
Adding Newsroom Management, Editor, Article to DApp

### PR DESCRIPTION
**Pull request by nickreynolds**
- Add Editor page which publishes article to IPFS and prints URL
- Add Create Newsroom page to deploy newsroom contract from DApp
- Add Newsroom Management page to add/view editors, reporters, articles
- Add Article page to render content by newsroom + articleId 
- Create TransactionButton component

note: the data is loaded into components in a pretty hacky way, but it's important to get this loop closed ASAP so we can start using the DApp.

### CODE REVIEW

File: packages/core/src/contracts/newsroom.ts [hunk 1]
```diff
@@ -124,6 +124,19 @@ export class Newsroom extends BaseWrapper<NewsroomContract> {
...
   }
 
+  public editors(): Observable<EthAddress> {
+    return this.instance
+      .RoleAddedStream({ role: Roles.Editor }, { fromBlock: 0 })
+      .map(e => e.args.grantee)
+      .concatFilter(async e => this.isEditor(e));
+  }
+
+  public reporters(): Observable<EthAddress> {
```
Comment (ritave) [inline packages/core/src/contracts/newsroom.ts]: I imagine we're removing this soon

File: packages/core/src/contracts/tcr/civilTCR.ts [hunk 2]
```diff
@@ -87,9 +108,85 @@ export class CivilTCR extends BaseWrapper<CivilTCRContract> {
...
-  /**
-   * Event Streams
-   */
+  //#region EventStreams
+
+  //#region RawEventStreams
+
+  public listingApplications(
+    listingAddress: EthAddress,
+  ): Observable<TimestampedEvent<DecodedLogEntryEvent<ApplicationArgs, string>>> {
```
Comment (ritave) [inline packages/core/src/contracts/tcr/civilTCR.ts]: Ops, forgot to add EventLogs

`DecodedLogEntryEvent<ApplicationArgs, string>` -> `ApplicationLogEvent`

int the next pr

File: packages/core/src/types.ts [hunk 3]
```diff
@@ -177,3 +179,38 @@ export interface ParamProp {
   proposedValue: BigNumber;
   pollID?: BigNumber;
 }
+
+export interface TimestampedEvent<T extends DecodedLogEntryEvent> extends DecodedLogEntryEvent {
+  timestamp(): Promise<number>;
+}
+
+export function createTimestampedEvent<T extends DecodedLogEntryEvent>(
```
Comment (ritave) [inline packages/core/src/types.ts]: move to utils, it gets exported outside

File: packages/core/src/types.ts [hunk 4]
```diff
@@ -177,3 +179,38 @@ export interface ParamProp {
...
+  web3Wrapper: Web3Wrapper,
+  event: T,
+): TimestampedEvent<T> {
+  return {
+    address: event.address,
```
Comment (ritave) [inline packages/core/src/types.ts]: Use destructing

```typescript
{
...event, // That's the syntax for this
timestamp: async () => ...
}
```

File: packages/core/src/types.ts [hunk 5]
```diff
@@ -81,6 +82,7 @@ export enum AbiType {
 }
 
 export { CivilLogs, Artifact } from "./contracts/generated/artifacts";
+export { ApplicationArgs, NewListingWhitelistedArgs } from "./contracts/generated/civil_t_c_r";
```
Comment (ritave) [inline packages/core/src/types.ts]: Really uncomfortable with this

File: packages/core/src/types.ts [hunk 6]
```diff
@@ -177,3 +179,38 @@ export interface ParamProp {
   proposedValue: BigNumber;
   pollID?: BigNumber;
 }
+
+export interface TimestampedEvent<T extends DecodedLogEntryEvent> extends DecodedLogEntryEvent {
```
Comment (ritave) [inline packages/core/src/types.ts]: `TimestampedEvent<T extends DecodedLogEntryEvent> extends T`
should work

File: packages/core/src/types.ts [hunk 7]
```diff
@@ -177,3 +179,38 @@ export interface ParamProp {
...
+  };
+}
+
+// TODO(ritave): generate roles from smart-contract
+/**
+ * Roles that are supported by the Newsroom
+ * - Editor can approve or deny contant, as well as assigning roles to actors
+ * - Reported who can propose content for the Editors to approve
+ */
+export enum Roles {
```
Comment (ritave) [inline packages/core/src/types.ts]: change name to something like `NewsroomRoles`

File: packages/dapp/src/components/Editor.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+  private onChange = (value: any): any => {
+    this.setState({ value });
+  };
+
+  private onProposeClicked = async (e: any): Promise<void> => {
+    this.setState({ publishInProgress: true });
+    const civil = getCivil();
+    const value = JSON.stringify(this.state.value.toJSON());
+    const uri = await civil.publishContent(value);
```
Comment (ritave) [inline packages/dapp/src/components/Editor.tsx]: Why not through the newsroom? Newsroom.propose() puts it too

File: packages/core/src/contracts/newsroom.ts (hunk 1 above)
Comment (nickreynolds) [reply packages/core/src/contracts/newsroom.ts]: I think we might still have reporters on the smart contract, it just makes sense to me. That way a reporter can both accept a position, and quit their position, if they want.

File: packages/core/src/types.ts (hunk 5 above)
Comment (nickreynolds) [reply packages/core/src/types.ts]: ok, I'll add a TODO to refactor that

File: packages/dapp/src/components/Editor.tsx (hunk 8 above)
Comment (nickreynolds) [reply packages/dapp/src/components/Editor.tsx]: this editor component doesn't assume anything about a newsroom. Walker and I were talking about the ContentProvider system in general and we think it needs a bit of a refactor. The DApp needs to interact with many different content providers at different times (based on user preference and/or URL scheme) which isn't really how it's setup right now.

File: packages/core/src/contracts/newsroom.ts (hunk 1 above)
Comment (walfly) [reply packages/core/src/contracts/newsroom.ts]: we should meet about this soon

File: packages/dapp/src/components/Article.tsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+class Article extends React.Component<ArticleProps, ArticleState> {
+  constructor(props: ArticleProps) {
+    super(props);
+    this.state = {
+      value: undefined,
+      error: "",
+    };
+  }
+
+  public async componentWillMount(): Promise<void> {
```
Comment (walfly) [inline packages/dapp/src/components/Article.tsx]: no wills,  use did mount.
https://reactjs.org/docs/react-component.html#unsafe_componentwillmount

File: packages/dapp/src/components/Article.tsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        {this.state.value && (
+          <CivilEditor plugins={plugins} value={this.state.value} onChange={this.onChange} readOnly={true} />
+        )}
+        {!this.state.value && "loading..."}
+      </>
+    );
+  }
+
+  private onChange = ({ value }: { value: any }): void => {
+    // gave warning (treated as error on production build) if function left empty
```
Comment (walfly) [inline packages/dapp/src/components/Article.tsx]: i guess thats my bad, should have default

File: packages/dapp/src/components/Article.tsx [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    return;
+  };
+
+  private initArticle = async () => {
+    const newsroom = await this.getNewsroom();
+    if (newsroom) {
+      const article = await newsroom.loadArticle(this.props.match.params.articleId);
```
Comment (walfly) [inline packages/dapp/src/components/Article.tsx]: i think ultimately the newsroom class needs to do less, it should just return the uri and your app needs to deal with it

File: packages/dapp/src/components/Editor.tsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  text-align: center;
+  &.active {
+    color: #30e8bd;
+  }
+  &.disabled {
+    background-color: #c4c2c0;
+  }
+`;
+
+export const initialJSON = {
```
Comment (walfly) [inline packages/dapp/src/components/Editor.tsx]: `import {defaultValue} from "@joincivil/editor";

File: packages/dapp/src/components/Editor.tsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
+import * as React from "react";
+import { CivilEditor, plugins } from "@joincivil/editor";
+import styled from "styled-components";
+import { getCivil } from "../helpers/civilInstance";
+
+// TODO(nickreynolds): get colors from constants file
+const Button = styled.button`
+  background-color: #2b56ff;
```
Comment (walfly) [inline packages/dapp/src/components/Editor.tsx]: should probably store civil colors as a constant somewhere

File: packages/dapp/src/components/Editor.tsx (hunk 8 above)
Comment (walfly) [reply packages/dapp/src/components/Editor.tsx]: i think publishing and retrieving content should be out of the core libraries scope

File: packages/dapp/src/components/Main.tsx [hunk 14]
```diff
@@ -1,21 +1,59 @@
...
+  civilSet: boolean;
+}
+class Main extends React.Component<{}, MainState> {
+  constructor(props: {}) {
+    super(props);
+    this.state = {
+      civilSet: false,
+    };
+  }
+  public componentWillMount(): void {
```
Comment (walfly) [inline packages/dapp/src/components/Main.tsx]: componentDidMount.  react is removing all Will methods to support async rendering

File: packages/dapp/src/components/Main.tsx [hunk 15]
```diff
@@ -1,21 +1,59 @@
...
+    window.addEventListener("load", this.initCivil);
+  }
+
+  public componentWillUnmount(): void {
```
Comment (walfly) [inline packages/dapp/src/components/Main.tsx]: except this one,  componentWillUnmount is still kosher

File: packages/dapp/src/components/listing/ListingHistory.tsx [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  constructor(props: any) {
+    super(props);
+    this.state = {
+      listingHistory: List<any>(),
+      compositeSubscription: new Subscription(),
+      error: undefined,
+    };
+  }
+
+  public async componentWillMount(): Promise<void> {
```
Comment (walfly) [inline packages/dapp/src/components/listing/ListingHistory.tsx]: did mount

File: packages/dapp/src/components/listing/ListingHistory.tsx [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  public componentWillUnmount(): void {
+    this.state.compositeSubscription.unsubscribe();
+  }
+
+  public render(): JSX.Element {
+    return (
+      <StyledDiv>
+        History:<br />
```
Comment (walfly) [inline packages/dapp/src/components/listing/ListingHistory.tsx]: br, your a mad man

File: packages/dapp/src/components/navbar/NavBar.tsx [hunk 18]
```diff
@@ -46,16 +46,22 @@ class NavBar extends React.Component<{}, NavBarState> {
           </NavBarLink>
         </NavBarItem>
         <NavBarItem>
-          <NavBarLink to="registry">Registry</NavBarLink>
+          <NavBarLink to="/registry">Registry</NavBarLink>
```
Comment (walfly) [inline packages/dapp/src/components/navbar/NavBar.tsx]: store the nav as json and use map?

File: packages/dapp/src/components/listing/ListingHistory.tsx [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    const newHistory = this.state.listingHistory.push({ ...event, timestamp });
+    this.setState({ listingHistory: newHistory.sort((a, b) => b.blockNumber - a.blockNumber).toList() });
+  };
+
+  // TODO(nickreynolds): move this all into redux
+  private initHistory = async () => {
+    const tcr = getTCR();
+
+    if (tcr) {
+      this.state.compositeSubscription.add(
```
Comment (walfly) [inline packages/dapp/src/components/listing/ListingHistory.tsx]: can you do this with some kind of iterable

File: packages/dapp/src/components/Listings.tsx [hunk 20]
```diff
@@ -21,47 +31,119 @@ class Listings extends React.Component<{}, ListingsState> {
...
-      });
-    }
+    const tcr = getTCR();
 
     if (tcr) {
-      const subscription = tcr.listingsInApplicationStage().subscribe(listing => {
-        this.setState({ applications: this.state.applications.push(listing) });
-      });
-      this.setState({ applicationSubscription: subscription });
+      this.state.compositeSubscription.add(
```
Comment (walfly) [inline packages/dapp/src/components/Listings.tsx]: feel like there should be  a better way

File: packages/dapp/src/components/newsroom/NewsroomManagement.tsx [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+      articleURL: "",
+      owners: [],
+      editors: List<string>(),
+      reporters: List<string>(),
+      proposedArticleIds: List<string>(),
+      compositeSubscription: new Subscription(),
+    };
+  }
+
+  public async componentWillMount(): Promise<void> {
```
Comment (walfly) [inline packages/dapp/src/components/newsroom/NewsroomManagement.tsx]: same

File: packages/dapp/src/components/utility/TransactionButton.tsx [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+        {this.state.error}
+        <Button onClick={this.onClick} disabled={this.state.disableButton}>
+          {this.state.disableButton && "Transaction Processing..."}
+          {!this.state.disableButton && this.props.children}
+        </Button>
+      </>
+    );
+  }
+
+  private onClick = async () => {
```
Comment (walfly) [inline packages/dapp/src/components/utility/TransactionButton.tsx]: does private work? react does some funky stuff with event handlers

**Comment by walfly** [comment]
componentWillMount is the only blocker

File: packages/dapp/src/components/Listings.tsx (hunk 20 above)
Comment (nickreynolds) [reply packages/dapp/src/components/Listings.tsx]: this subscription stuff is all going to change

File: packages/dapp/src/components/utility/TransactionButton.tsx (hunk 22 above)
Comment (nickreynolds) [reply packages/dapp/src/components/utility/TransactionButton.tsx]: the button works (doesn't do what I want it to do on disable though...)

**Pull request merged**